### PR TITLE
Add RTPS publisher and subscribers

### PR DESCRIPTION
**Pull request by lpereira**
This is a PoC that publishes data using the RTPS protocol, using the Fast-RTPS and Fast-CDR libraries.  These are the same libraries used by the ROS project.

These libraries require code to be built with C++ exceptions, which isn't possible with ArduCopter at the moment.  For this reason, AP_RTPS has been introduced: this library has all the exception handling gunk and wraps those as return values.  This seems safe as that's the same approach used to use C++ libraries with C code.

Most of the this library has been adapted from code generated by the fastrtpsgen tool provided by the Fast-RTPS libraries, but made generic by using C++ templates.  This has advantages and disadvantages, but I believe the advantages outweigh the disadvantages, given that this significantly reduces the amount of boilerplate code necessary to publish a different kind of topic.

One still has to copy+edit a few of the generated files, though, isn't optimal.  The AHRS.idl file is provided as documentation only, since the generator is not called during build time as it should be done in an ideal world.

Also, please note the license: the code is under the same terms as the generated code is (Apache 2.0), but the license is compatible with the rest of the project (GPLv3).

Explanation on how to test and add new messages is available under libraries/AP_RTPS/README.

### CODE REVIEW

File: ArduCopter/ArduCopter.cpp [hunk 1]
```diff
@@ -152,6 +152,10 @@ const AP_Scheduler::Task Copter::scheduler_tasks[] = {
     SCHED_TASK(userhook_SuperSlowLoop, 1,   75),
 #endif
     SCHED_TASK(button_update,          5,    100),
+#if defined(HAVE_FASTRTPS) && defined(HAVE_FASTCDR) && \
```
Comment (magicrub) [inline ArduCopter/ArduCopter.cpp]: for clarity, would be nice for this to be on one line

File: libraries/AP_RTPS/AHRS.h [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+#include <string>
+#include <vector>
+
+namespace eprosima {
+namespace fastcdr {
+    class Cdr;
+}
+}
+
+class AHRS {
```
Comment (magicrub) [inline libraries/AP_RTPS/AHRS.h]: this file and class need a new name

**Comment by OXINARF** [comment]
Was this talked about somewhere and I missed it? The only thing I remember about this was 3DR talking about it and saying they would implement it in Solo and we saying we didn't see great benefit.
I mean, this is quite a change and I'm sure we aren't going to take this in just as a PoC.

File: libraries/AP_RTPS/AHRS.h (hunk 2 above)
Comment (lpereira) [inline libraries/AP_RTPS/AHRS.h]: Would it be better if it were under a `rtps` namespace instead of changing the name?

**Comment by rmackay9** [comment]
I guess my question is the same as @OXINARF's.  It's certainly very interesting but I'm missing what the end goal is.  If it helps with our integration with ROS then I'm definitely keen.  .. or perhaps it's to allow a better/smoother integration with a companion computer instead of using MAVLink?  That could also be quite valuable.

File: libraries/AP_RTPS/AHRS.h (hunk 2 above)
Comment (magicrub) [inline libraries/AP_RTPS/AHRS.h]: if anything, just call it RTPS_AHRS. Goal is dis-ambiguity.

**Comment by lucasdemarchi** [comment]
@rmackay9 Yes, the intention is to allow better integration with a companion computer (or in case of Linux it might be the same computer). RTSP was chosen due to ros2 to allow it to be used without a bridge.

Tully Foote talked about it during Dronecode Unconference this year. We should probably get his opinion on this. @vmayoral, you may also be interested.

I'm disappointed a little bit about the generated code.... It doesn't seem as lightweight as people say. Anyway this is a proposal for discussion

**Comment by lucasdemarchi** [comment]
@oxinarf I forgot to mention you on the previous comment as well.

**Comment by vmayoral** [comment]
Definitely interested on this. Thanks for bringing it up @lpereira!

Having native RTPS pub/sub capabilities would allow the autopilots to be first class participants of the ROS 2 ecosystem. This will have several advantages for the people using ROS 2 and as @lucasdemarchi points out, even bridges like `mavros` could be removed. 

Furthermore, this will open up our autopilots to the whole ecosystem of ROS sensors/actuators. That is, simply by creating an abstraction that subscribes to the appropriate RTPS topic, we could easily support new cameras, depth sensors, range finders that already are well defined in the ROS world. Simulation with Gazebo will also benefit  (right now the architecture for a proper simulation is a bit complex).

>One still has to copy+edit a few of the generated files, though, isn't optimal. The AHRS.idl file is provided as documentation only, since the generator is not called during build time as it should be done in an ideal world.

We can probably benefit from the ROS 2 tooling. @lpereira, have you considered using the ROS 2 primitives instead of the (Fast)RTPS ones?

There're several reasons why this could be interesting, the most relevant one is to avoid vendor locking. FastRTPS belongs to eProsima, which is a DDS/RTPS vendor doing a good job but there're many other interesting initiatives, e.g.: http://github.com/ros2/freertps could potentially lead to FastRTPS integration in the popular STM32FX family. If we were to code on top of the ROS 2 primitives, we could easily allow people to switch between DDS vendors depending on their interests (e.g.: RTI is probably the most popular DDS vendor, widely accepted on industrial/military applications) and platforms.

**Comment by magicrub** [comment]
Sounds like a winner to me, thanks for this!

**Comment by tridge** [comment]
yes, the big win is ROS2 integration. Thanks very much for your work on this PoC !

**Comment by OXINARF** [comment]
@lucasdemarchi @vmayoral Thanks for your posts, it helps understanding the need for this.

I think that we need to thing of a better integration though. We have MAVLink, we have FrSky and now we want to have RTPS. We should really consider an AP_GCS library with those as backends. I think the hard part about it would be receiving commands (because vehicles use MAVLink messages directly and differently).

**Comment by tridge** [comment]
@OXINARF yes, I agree a better approach would be nice
It is mostly about state information, and the hardest part of state information that is not already stored in libraries, such as state in vehicle specific code.
I think a fairly simple approach would be to make the current XXX_send() functions we call for MAVLink send calls instead fill in elements of a state structure owned by libraries/GCS_State. That state structure could then be used by all protocols. We would not include in the state structure information which is already readily accessible via existing classes like AHRS. I think the state structure would be quite small (maybe 100 bytes or so?) but it would make doing lots of different protocols quite easy.

**Comment by lucasdemarchi** [comment]
@tfoote this may be of your interest.  One of the goals is to be easy integration of ROS2 with ardupilot.  Consider this PR as PoC so we can get things going.

Could you share an update on where you are in ROS with regard your talk on ELC this year.... we may need some more work to catch up on your development.

**Comment by lvale** [comment]
IMHO, this is a very important development. Great work Leandro @lpereira I really hope for your further developments on top of this great work.

**Comment by tfoote** [comment]
@lucasdemarchi Thanks for the ping. 

We've made some good progress on developing a standard interface. We just updated [REP 105](http://www.ros.org/reps/rep-0105.html) for better outdoor localization support needed for drones. And [REP 147](http://www.ros.org/reps/rep-0147.html) is drafted and [under review](https://github.com/ros-infrastructure/rep/pull/120)

FastRTPS is one of the main implementations we're using in our ROS2 prototypes. It should be a small step to replace the mavlink to ROS bridge with either a FastRTPS compatibility layer or possibly extend the built in FastRTPS layer to send messages with the extra metadata to integrate with a ROS2 system. If things go very well we may even be able to support compiling full ROS2 support into the autopilot. We're working hard to keep the footprint low with the goal of being able to compile onto microcontrollers. 

The idea from @OXINARF to have an abstraction of the transport would be a good approach. As @vmayoral pointed out providing an interface that is transport independent will be valuable too. 

@lpereira Have you measured how this affects the installation footprint/memory usage?

**Comment by lpereira** [comment]
Everyone: thanks for all the kind comments. Answering some of the questions that have been directly addressed to me and not answered by @lucasdemarchi already:

@vmayoral: I did, yes, but there is a big fat warning in its wiki suggesting that it's an experimental piece of code. So I chose to use a more mature implementation instead, even if I dislike some of its design decisions. Given that the surface touching the protocol is a small one, moving to another implementation should be possible without too much hassle.

@tfoote: I haven't measured memory consumption and installation footprint. I've just ran `size` on some of the FastRTPS library and some of its dependencies here (Linux, x86-64), just to get an idea of the installation footprint:

```
   text       data        bss        dec        hex    filename
1401575      34120       4536    1440231     15f9e7    libfastrtps.so
  54222       1560          8      55790       d9ee    libfastcdr.so
 156255       7024        232     163511      27eb7    /lib64/libboost_thread.so.1.60.0
   8849       1032        128      10009       2719    /lib64/libboost_system.so.1.60.0
  58883       3776         88      62747       f51b    /lib64/libboost_date_time.so.1.60.0
  20072       1568          8      21648       5490    /lib64/libboost_chrono.so.1.60.0
1067914      25352        736    1094002     10b172    /lib64/libboost_regex.so.1.60.0
   2102        608       2688       5398       1516    /lib64/libboost_atomic.so.1.60.0
25046121        560          8    25046689    17e2ea1    /lib64/libicudata.so.56
2534694      59758       3440    2597892     27a404    /lib64/libicui18n.so.56
1592862      65824      10784    1669470     19795e    /lib64/libicuuc.so.56
```

The added footprint is quite large, specially if considering the dependency on libicu (which is a requirement for libboost-regex).

The disk footprint difference between the flight stack with and without RTPS support should be negligible right now, but can spiral up quite quickly due to reliance on templated code (and having many instances for each kind of message). Since this is a proof-of-concept, I'm just avoiding any kind of size reduction.

**Comment by magicrub** [comment]
This has stalled. We want it, but is this how we want it?

**Comment by lucasdemarchi** [comment]
Closing as we will need to rework it to get into a state to be applied.

**Pull request closed without merging**